timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockWatchingServiceImpl.java

```diff
@@ -193,15 +193,17 @@ private static Optional<LockRequestMetadata> filterMetadataBasedOnFilteredLocks(
             Optional<LockRequestMetadata> unfilteredMetadata) {
         return unfilteredMetadata
                 .map(LockRequestMetadata::lockDescriptorToChangeMetadata)
-                .map(unfilteredLockMetadata -> {
+                .flatMap(unfilteredLockMetadata -> {
                     Map<LockDescriptor, ChangeMetadata> filteredLockMetadata = KeyedStream.ofEntries(
                                     unfilteredLockMetadata.entrySet().stream())
                             .filterKeys(lockDescriptor -> {
                                 assert unfilteredLocks.contains(lockDescriptor) : "Unknown lock descriptor in metadata";
                                 return filteredLocks.contains(lockDescriptor);
                             })
                             .collectToMap();
-                    return LockRequestMetadata.of(filteredLockMetadata);
+                    return filteredLockMetadata.isEmpty()
+                            ? Optional.empty()
+                            : Optional.of(LockRequestMetadata.of(filteredLockMetadata));
```

Comment:
This is no necessary for correctness, but we should have this behavior for performance reasons

Reply:
Rather than ternary I personally bias towards 
``` 
if(filteredLockMetadata.isEmpty()) {
   return Optional.empty();`
}
return Optional.of(...) 
```

BUT -- that said, I think ternary is also fine!

Reply:
I can get behind it